R/pkg/R/deserialize.R

```diff
@@ -197,6 +197,31 @@ readMultipleObjects <- function(inputCon) {
   data # this is a list of named lists now
 }
 
+readMultipleObjectsWithKeys <- function(inputCon) {
+  # readMultipleObjectsWithKeys will read multiple continuous objects from
+  # a DataOutputStream. There is no preceding field telling the count
+  # of the objects, so the number of objects varies, we try to read
+  # all objects in a loop until the end of the stream. The rows in
+  # the stream are separated by grouping-key boundary
+  data <- list()
+  subData <- list()
+  while (TRUE) {
+    # If reaching the end of the stream, type returned should be "".
+    type <- readType(inputCon)
+    if (type == "") {
+      break
+    } else if (type == "r") {
+      # key boundary detected
+      readTypedObject(inputCon, type)
```

Comment:
it is actually a boundary separator not the key we used to refer in : https://github.com/NarineK/spark/blob/9cacd4dbfa0e20d2a855e23f2962a258abbba553/R/pkg/inst/worker/worker.R#L94

No, it is not being used internally, it is a separator(to mark the group boundaries)